### PR TITLE
Mgs conanize

### PR DESCRIPTION
**Pull request by martukas**
This pull request fully conanizes the EFU, and also includes a thorough refactoring of the CMake files. These are the most significant changes:

* Everything builds with conan

* Multi-platform build

* All binaries end up in `build_root/bin`

* All libraries end up in `build_root/lib`

* The `efu2` is now called simply `efu`. No one remembers an efu1 so the designation is meaningless. We will do proper semantic versioning at some point. All references to efu2 have been replaced in documentation, etc.. I guess you will need to update your grafana screens and deployment scripts.

* In view of the above 3 points, the way to run gdgem would be (in build root):

``./bin/efu -d ./lib/gdgem -f ~/cfg.json``

* There is no more local flatbuffer headers. They are all brought in via conan and added to the path.

* There is no more local ESSGeometry. This is also brought in via conan.

* The implication of the above 2 points is that, essentially, conan builds are the only ones we now explicitly support. That is how all the docker nodes do it on Jenkins, and it is enough work to keep everything compatible via this mechanism. If you want to somehow build without conan, that should still be possible by manually setting some include path variables when in voking CMake. I think? But if supporting it demands some additional complexity in the CMake scripts, we will avoid this. We have tried to keep support for both ways on the file writer at PSI's behest, but it has just ground everything to a halt. This is partly why I have gone crazy on the EFU this week, while spending boring hours waiting for PSI to confirm that the file writer builds without conan on some even older Centos that we have never seen nor have any Jenkins nodes for...

* I have refactored the CMake stuff almost unrecognizably. Everything should work, but there are some outstanding questions. If you search through the files for a tripple-question mark "???" you will find some comments.

* FYI, when building, on some systems you need to invoke the conan-provided virtual environment. So the new mantra goes like:
```
mkdir build

cd build

conan install --build=outdated ..

source activate_run.sh

cmake -DUSE_OLD_ABI=0 -DDUMPTOFILE=1 ..

make
```

### CODE REVIEW

**Comment by SkyToGround** [comment]
I have started looking at this now. Please give me until the end of the day (at the very least) before merging the branch.

**Comment by mortenjc** [comment]
Please don't merge this to master before I have succeedd in

building this on my mac!

File: .gitignore [hunk 1]
```diff
@@ -11,7 +11,7 @@
 cncsread
 batchreader
 genpixids
-efu2
+efu
```
Comment (SkyToGround) [inline .gitignore]: 👍

File: .gitignore [hunk 2]
```diff
@@ -20,7 +20,7 @@ dgmgjson
 *Test$
 
 #directories
-build/
+build*/
```
Comment (SkyToGround) [inline .gitignore]: 👍

**Comment by SkyToGround** [comment]
I had some build issues on my machine that needs to be resolved. In my opinion, this branch should NOT be merged at this point in time. I will get back to you on fixes or suggested fixes for those issues.

**Comment by SkyToGround** [comment]
The README.md file must be updated to include working instructions on how to build the efu.

**Comment by SkyToGround** [comment]
There is at least one conan package missing: _boost_.

**Comment by martukas** [comment]
Boost comes in transitively via h5cpp. It does not need to be repeated, at the risk of actually creating a conflict if it differs from h5cpp's version.

**Comment by SkyToGround** [comment]
I had issues with using an older version of conan and the correct version of cmake not being selected. The former should be documented and the latter should be fixed.

**Comment by SkyToGround** [comment]
With some help from Afonso we determined that boost should have been provided via h5cpp but that part did not work on the Ubuntu system that I was testing on.

File: prototype2/Utilities.cmake [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    ${${module_name}_SRC}
+    ${${module_name}_INC})
+  set_target_properties(${module_name} PROPERTIES PREFIX "")
+  set_target_properties(${module_name} PROPERTIES SUFFIX ".so")
+  target_link_libraries(${module_name}
+    ${${module_name}_LIB}
+    ${EFU_COMMON_LIBS}
+    eventlib)
+  enable_coverage(${module_name})
+  install(TARGETS ${module_name} DESTINATION bin)
```
Comment (SkyToGround) [inline prototype2/Utilities.cmake]: This line is ignored (at least under MacOSX) for some reason. I have made a bunch of different attempts at fixing this we no success so far.

Comment (martukas) [reply prototype2/Utilities.cmake]: Which line specifically?
It has worked for me on my mac
We might also consider something like [this](https://github.com/ess-dmsc/h5cpp/blob/master/cmake/OutputDirConfig.cmake)
to standardize these paths.

**Comment by SkyToGround** [comment]
libpcap is a dependency that is missing from the conanfile. Do we want to provide that library as well or do we not care. CMake will find a "system version" of libpcap if available. Care to comment @mortenjc ?

**Comment by SkyToGround** [comment]
To keep everyone updated (@martukas , @mortenjc ): I am making good progress and my hope is that I will be finished with the tests and modifications of the build system by the end of tomorrow.

**Comment by SkyToGround** [comment]
@martukas The *logical-geometry* dependency is a problem in my opinion. When compiling the EFU without using conan to provide the dependencies, you have to manually provide the location of the `logical_geometry/ESSGeometry.h` file. Problem is that this directory is only created by the conan package for logical-geometry. There are several solutions to this, some of which are listed below:

1. Modify the repository of *logical-geometry* to reflect what the include statement in `mbcaen.cpp`  expects.
2. Modify the include statement in `mbcaen.cpp` to reflect the actual location of `ESSGeometry.h` in the *logical-geometry* repository.
3. Modify the build system of the *event-formation-unit* to download and move `ESSGeometry.h`.
4. Put back `ESSGeometry.h` in the *event-formation-unit* repository.
5. Modify the *event-formation-unit* build system to not build the *mbcaen* detector module of the `ESSGeometry.h` file is not present.

I would personally prefer either 4 or 5. Do you have an opinion @mortenjc ?

**Comment by SkyToGround** [comment]
I have now finished testing of building the EFU both manually and using conan under MacOSX, Ubuntu 16 and (conan only) Centos 7.4. There are a few things that needs to be fixed before I think we can merge this branch with master.

**Comment by mortenjc** [comment]
ESSGeometry: hmm. Is this any different than for the streaming-data-types? We need to share header files and ideally we should not duplicate them. So basically ESSGeometry needs to be installed for the EFU to compile so,...?.

**Comment by SkyToGround** [comment]
@mortenjc I discussed this with @martukas yesterday and as I told him; seeing as the EFU is currently the only project using the logical-geometry repository and it only contains one useful file (ESSGeometry.h) I thought that the repository was not that useful. However, it was explained to me that a more code will be moved to the repository in the near future. Thus option 4 is probably not a very good one.

Based on discussions with @amues , it seems to me that the better option is to re-arrange the structure of the logical-geometry repository.

**Comment by SkyToGround** [comment]
After talking to @mortenjc , I am now merging the branch to master.

**Pull request merged**